File: 49-group-anagrams/group-anagrams.java
class Solution {
    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> groups = new HashMap<>();
        int n = strs.length;
        
        for(int i = 0; i < n; i++) {
            String key = "";

            Map<Character, Integer> count = new HashMap<>();
            String str = strs[i];
            for (char c = 'a'; c <= 'z'; c++) {
                count.put(c, 0); // Initialize with a default value of 0
            }

            for (int j = 0; j < str.length(); j++) {
                Character c = (Character) str.charAt(j);
                count.put(c, count.getOrDefault(c, 0) + 1);
            }

            for (Map.Entry<Character, Integer> entry: count.entrySet()) {
                key += (entry.getKey() + ":" + entry.getValue());
            }
          
            groups.put(key, groups.getOrDefault(key, new ArrayList<>()));
            groups.get(key).add(str);
        }

        List<List<String>> result = new ArrayList<>();

        for (Map.Entry<String, List<String>> entry : groups.entrySet()) {
            result.add(entry.getValue());
        }

        return result;
    }
}

File: 49-group-anagrams/group-anagrams.swift
class Solution {
    func groupAnagrams(_ strs: [String]) -> [[String]] {
        var a = Character("a").asciiValue
        var groups:[String:[String]] = [:]

        for s in strs {
            var counts: [Int] = Array(repeating: 0, count: 26)

            for (i, c) in s.enumerated() {
                var idx = Int(c.asciiValue! - a!)
                counts[idx] += 1
            }
            var key: String = ""
            for (i, v) in counts.enumerated() {
                if v > 0 {
                    key += "\(i):\(v)"
                }
            }
            groups[key, default: []].append(s)
        }

        return Array(groups.values)
    }
}